LPC55S69 board with TFM samples

agansari opened the pull request:
Builds lpcxpresso55s69_ns target with TFM samples.

Limitations:
 - builds without MCUboot
 - flashing is done manually

Depends on: https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/12
Used #27524 as inspiration

agansari committed:
samples: tfm_integration: BL2 configuration

Configure build variable BL2 depending on board configuration. BL2 is
set to True by default. BL2 enabled building TFM with MCUboot.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
boards: lpcxpresso55s69 mcuboot and tfm partitions

Moved the flash partitioning to the shared device tree include so all
targets can benefit from the partitioning.
This is implemented to MCUboot can run TFM(S) with Zephyr
(lpc55s69 cpu 0 NS).
Also defines RAM partitions as TFM expects them, first part of ram for
S, second half for NS.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
boards: lpcxpress55s69: NS target with TFM support

Configures the Non-Secure target to work with TFM.
Memory (Flash and RAM) mapping according to TFM definitions.
How to build and flash the board.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>
Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

agansari committed:
samples: tfm_integration: lpc55s69 support

Adds documentation on how to run TFM samples on LPCxpresso55S69 board.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

d3zd3z commented on boards/arm/lpcxpresso55s69/lpcxpresso55s69.dtsi:
```diff
@@ -90,6 +90,64 @@
+		};
+		scratch_partition: partition@60000 {
+			label = "image-scratch";
+			reg = <0x00060000 0x0003d800>;
```
There isn't much point in making scratch this large. We have a rather large flash device, it would be better to make the two slots somewhat less than half, and make the scratch a few sectors at most (there are wear issues if scratch is too small). Realistically, we should be configuring `BOOT_SWAP_USING_MOVE` and not have a scratch partition. This requires that slot 0 be only a single sector larger.

d3zd3z commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -9,3 +10,86 @@ if(CONFIG_PINMUX_MCUX_LPC)
+	# 		-offset 0x0
+	# 		-o ${CMAKE_BINARY_DIR}/tfm_image.hex -intel --line-length=44
+	# )
+endif()
```
This file is missing a newline at the end.

d3zd3z commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -9,3 +10,86 @@ if(CONFIG_PINMUX_MCUX_LPC)
+		message(FATAL_ERROR "'srec_cat' not found. Please install it, or add it to $PATH.")
+	endif()
+
+	# MCUboot, TFM and Zephyr intergration is not yet correctly implemented
```
Should we be committing commented out code? Or can this be added later when it is working. If you want to make the code available for others to look at, it could be put in its own pull request (marked as draft if it isn't working), without the comments.

d3zd3z commented on boards/arm/lpcxpresso55s69/lpcxpresso55s69_ns.dts:
```diff
@@ -24,8 +24,9 @@
-		zephyr,sram = &sram0;
+		zephyr,sram = &non_secure_ram;
 		zephyr,flash = &flash0;
+		zephyr,code-partition = &slot1_partition;
```
If we intend to support upgrades (which should work at least with the Zephyr driver), we'll also need slots 2 and 3 to hold the upgrade images of the two images.

agansari commented on samples/tfm_integration/psa_level_1/CMakeLists.txt:
```diff
@@ -7,12 +7,19 @@ set(QEMU_KERNEL_OPTION "-device;loader,file=${CMAKE_BINARY_DIR}/tfm_qemu.hex")
 
 find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
 
+set(BL2 $ENV{BL2})
+if(NOT DEFINED BL2)
+    #by default build with MCUboot
+    set(BL2 True)
+endif()
```
> I'm OK with this, but it will require updating the `mps2_an521_nonsecure` and `nucleo_l552ze_q` to set the BL2 status to true:
> 
> * https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/tfm_integration/tfm_ipc/CMakeLists.txt#L11
> * https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/tfm_integration/psa_level_1/CMakeLists.txt#L10

@microbuilder @tejlmand i've first set BL2 default True here and the other TFM sample.

tejlmand replied:
This can be removed, in TFM the BL2 is True when using `CFGFILE "ConfigRegressionIPC"` as is done in this file:
https://github.com/zephyrproject-rtos/zephyr/blob/7e70b77a78d3c565596b0bef11878748d7cfdff9/samples/tfm_integration/psa_level_1/CMakeLists.txt#L20

https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/master/trusted-firmware-m/configs/ConfigRegressionIPC.cmake#L59-L61

Note: depends on this PR comment: https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/12#discussion_r481984582

agansari replied:
@tejlmand thank you, i've removed this change, tested with zephyrproject-rtos/trusted-firmware-m#12 (comment)

agansari commented:
@d3zd3z thank you for your review, can you please review again?

microbuilder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -126,6 +125,24 @@ and run it in qemu via the ``run`` command.
+   .. code-block:: bash
+      $ west build -p -b lpcxpresso55s69_ns samples/tfm_integration/psa_level_1/ --
+
+For we build TFM wihout a bootloader (MCUboot), to run TFM images we need to flash them manually.
+
+   .. code-block:: console
+
+      JLinkExe -device lpc55s69 -if swd -speed 2000 -autoconnect 1
+      J-Link>loadfile build/tfm/install/outputs/LPC55S69/tfm_s.hex
+      J-Link>loadfile build/zephyr/zephyr.hex
```
```suggestion
Next we need to manually flash the secure (`tfm_s.hex`) and non-secure (`zephyr.hex`) images wth a J-Link as follows:

   .. code-block:: console

      JLinkExe -device lpc55s69 -if swd -speed 2000 -autoconnect 1
      J-Link>loadfile build/tfm/install/outputs/LPC55S69/tfm_s.hex
      J-Link>loadfile build/zephyr/zephyr.hex

NOTE: At present, the LPC55S69 doesn't include support for the MCUBoot bootloader.
```

microbuilder commented on samples/tfm_integration/tfm_ipc/README.rst:
```diff
@@ -131,6 +131,23 @@ Reset the board.
+   .. code-block:: bash
+      $ west build -p -b lpcxpresso55s69_ns samples/tfm_integration/tfm_ipc/ --
+
+For we build TFM wihout a bootloader (MCUboot), to run TFM images we need to flash them manually.
+
+   .. code-block:: console
+
+      JLinkExe -device lpc55s69 -if swd -speed 2000 -autoconnect 1
+      J-Link>loadfile build/tfm/install/outputs/LPC55S69/tfm_s.hex
+      J-Link>loadfile build/zephyr/zephyr.hex
```
```suggestion
Next we need to manually flash the secure (`tfm_s.hex`)  and non-secure (`zephyr.hex`) images wth a J-Link as follows:

   .. code-block:: console

      JLinkExe -device lpc55s69 -if swd -speed 2000 -autoconnect 1
      J-Link>loadfile build/tfm/install/outputs/LPC55S69/tfm_s.hex
      J-Link>loadfile build/zephyr/zephyr.hex

NOTE: At present, the LPC55S69 doesn't include support for the MCUBoot bootloader.
```

microbuilder commented on boards/arm/lpcxpresso55s69/doc/index.rst:
```diff
@@ -191,6 +195,20 @@ see the following message in the terminal:
+Building and flashing secure/non-secure with Arm |reg| TrustZone |reg|
+---------------------------------------------------------------------
+The TF-M integration samples can be run by a LPC55S69, using the ``lpcxpresso55s69_ns`` target.
+For we build TFM wihout a bootloader (MCUboot), to run TFM images we need to flash them manually.
```
This sentence is unclear. See the other comments to possibly clarify.

agansari replied:
@microbuilder please check this documentation again?

microbuilder reviewed (changes requested):
Shouldn't the west.yaml file reference to TF-M but updated to https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/12, and then the final hash when that PR is merged?

agansari committed:
west.yml: update to 143df67

Updates west.yml to use modules/tee/tfm commit number
143df675557305b61f7930a50459a53a8d2bb097.
Change: make BL2 configurable.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari commented:
> Shouldn't the west.yaml file reference to TF-M but updated to [zephyrproject-rtos/trusted-firmware-m#12](https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/12), and then the final hash when that PR is merged?

You are right, I've added another commit to update tfm module, also fixed some sanity checks.

microbuilder reviewed (changes requested):
The `tfm_ipc` and `psa_level_1` sample's `CMakeLists.txt` files will both require an addititon to set BL2 to false when calling `trusted_firmware_build`. The currently fail with this error since BL2 is set to TRUE by default:

```
$ west build -p -b lpcxpresso55s69_ns zephyr/samples/tfm_integration/psa_level_1/ --
...
[15/170] Performing configure step for 'tfm'
FAILED: tfm-prefix/src/tfm-stamp/tfm-configure 
cd /Users/kevin/Linaro/zephyr/upstream/build/tfm && /usr/local/Cellar/cmake/3.18.1/bin/cmake -DPROJ_CONFIG=/Users/kevin/Linaro/zephyr/upstream/modules/tee/tfm/trusted-firmware-m/configs/ConfigRegressionIPC.cmake -DTARGET_PLATFORM=LPC55S69 -DCOMPILER=GNUARM -DGNUARM_PREFIX=arm-none-eabi -DGNUARM_PATH=/usr/local/prog/gcc-arm-none-eabi-7-2018-q2-update -GNinja /Users/kevin/Linaro/zephyr/upstream/modules/tee/tfm/trusted-firmware-m && /usr/local/Cellar/cmake/3.18.1/bin/cmake -E touch /Users/kevin/Linaro/zephyr/upstream/build/tfm-prefix/src/tfm-stamp/tfm-configure
-- embedded_project_start: using project specific config file (PROJ_CONFIG = /Users/kevin/Linaro/zephyr/upstream/modules/tee/tfm/trusted-firmware-m/configs/ConfigRegressionIPC.cmake)
CMake Error at platform/ext/lpc55s69.cmake:53 (message):
  ERROR: BL2 is currently not supported on target 'LPC55S69'.
```

tejlmand commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -9,3 +10,29 @@ if(CONFIG_PINMUX_MCUX_LPC)
+		set(TFM_IMAGE_VERSION_NS 0.0.0+0)
+	endif()
+
+	set(ENV{BL2} False)
```
Why setting this as env variable ?
Should this be a Kconfig setting, so that `CONFIG_TFM_BL2` could be used instead ?

Then boards can specify their value for example in: https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0_defconfig

agansari replied:
will change this, samples will set BL2 False for this board.

tejlmand replied:
This could be something like:
```
choice TFM_BL2
        prompt "BL2"
        default TFM_BL2_CONFIG_FILE_DEFAULT

config TFM_BL2_CONFIG_FILE_DEFAULT
        bool "Use TFM BL2 setting from TFM configuration file"

config TFM_BL2_TRUE
        bool "TFM BL2 enabled"

config TFM_BL2_FALSE
        bool "TFM BL2 disabled"

endchoice
```
so that default is to use the TFM BL2 setting from conf file, or specifically enable/disable this feature.

microbuilder replied:
This would be a better approach, and if the default is `TFM_BL2_CONFIG_FILE_DEFAULT` we shouldn't need to touch the `mps2_an521` and `nucleo_l552ze_q` targets.

agansari replied:
@tejlmand @microbuilder made the change requested, please review how this is treated at board level and samples.

tejlmand commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -9,3 +10,29 @@ if(CONFIG_PINMUX_MCUX_LPC)
+	set(ENV{BL2} False)
+
+	set(PREPROCESSED_FILE "${CMAKE_BINARY_DIR}/tfm/image_macros_preprocessed")
+	set(TFM_MCUBOOT_DIR "${ZEPHYR_BASE}/../modules/tee/tfm/trusted-firmware-m/bl2/ext/mcuboot")
```
```suggestion
 set(TFM_MCUBOOT_DIR "${ZEPHYR_TFM_MODULE_DIR}/trusted-firmware-m/bl2/ext/mcuboot")
```

agansari replied:
agree

agansari commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -9,3 +10,29 @@ if(CONFIG_PINMUX_MCUX_LPC)
+
+	# Configure which format (full or hash) to include the public key in
+	# the image manifest
+	set(TFM_PUBLIC_KEY_FORMAT "full")
```
All other boards configure TFM_PUBLIC_KEY_FORMAT this way, maybe refactor this later?

tejlmand replied:
accepted.

agansari committed:
modules: tfm BL2 configurable

Make BL2 setting configurable via Kconfig choice.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
samples: tfm_integration: BL2 configurable

Make BL2 configurable via Kconfig.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

tejlmand reviewed (changes requested):
Some changes required to ensure that one can default to use the standard TFM config file value.

agansari commented on samples/tfm_integration/psa_level_1/CMakeLists.txt:
```diff
@@ -7,12 +7,19 @@ set(QEMU_KERNEL_OPTION "-device;loader,file=${CMAKE_BINARY_DIR}/tfm_qemu.hex")
                        IPC
                        CFGFILE "ConfigRegressionIPC"
                        OUT_VENEERS_FILE VENEERS_FILE
+                       BL2 ${TFM_BL2}
```
thank you for you suggestions @tejlmand i've pushed this changes

tejlmand reviewed (approved):
lgtm

MaureenHelm commented:
@d3zd3z please revisit

tejlmand commented on modules/Kconfig.tfm:
```diff
@@ -38,3 +38,18 @@ config TFM_KEY_FILE_NS
 	  The path and filename for the .pem file containing the private key
 	  that should be used by the BL2 bootloader when signing non-secure
 	  firmware images.
+
```
The Zephyr module menu starts to look quite messy, could you please apply the following patch.
```
diff --git a/modules/Kconfig.tfm b/modules/Kconfig.tfm
index a15f092da1..4d40d275d8 100644
--- a/modules/Kconfig.tfm
+++ b/modules/Kconfig.tfm
@@ -3,7 +3,7 @@
 # Copyright (c) 2019, 2020 Linaro Limited
 # SPDX-License-Identifier: Apache-2.0
 
-config BUILD_WITH_TFM
+menuconfig BUILD_WITH_TFM
        bool "Build with TF-M as the Secure Execution Environment"
        depends on TRUSTED_EXECUTION_NONSECURE
        help
@@ -21,6 +21,8 @@ config BUILD_WITH_TFM
            "mps2_an521_nonsecure") ensures that
            CONFIG_TRUSTED_EXECUTION_NONSECURE ie enabled.
 
+if BUILD_WITH_TFM
+
 config TFM_KEY_FILE_S
        string "Path to private key used to sign secure firmware images."
        depends on BUILD_WITH_TFM
@@ -53,3 +55,5 @@ config TFM_BL2_FALSE
         bool "TFM BL2 disabled"
 
 endchoice
+
+endif # BUILD_WITH_TFM
```

That will make TF-M related settings to have it's own menu in menuconfig looking like:
![image](https://user-images.githubusercontent.com/9802583/92218357-48fa6d80-ee99-11ea-8334-87a080ea79ae.png)
![image](https://user-images.githubusercontent.com/9802583/92218407-5ca5d400-ee99-11ea-8ddd-092368ef2645.png)

instead of:
![image](https://user-images.githubusercontent.com/9802583/92218464-7810df00-ee99-11ea-834c-73be0489692d.png)

agansari replied:
i've applied this patch

tejlmand reviewed (changes requested):
Please group TF-M and BL2 settings inside a TF-M menu.
